### ISSUE TITLE
siamese_contrastive.py save and load model?

### ISSUE THREAD

**Issue by sherlock666** [keras-team/keras-io#668]
According to the demo code 
"Image similarity estimation using a Siamese Network with a contrastive loss"
[https://keras.io/examples/vision/siamese_contrastive/](url)

I'm trying to save model by model.save  to h5 or hdf5
however after I used load_model (even tried load_weights)
it showed error message for : unknown opcode

Have done googling job which all tells me it's python version problem between py3.5~py3.6
But actually I use only python 3.8....
other info say that there's some extra job need to be done either in model building or load_model

It would be very kind for Keras team or any one to help provide the save and load model part
to make this demo code more complete
thanks!!

**Comment by anukash** [comment]
Hi sherlock666,
Actually here they are using two individual factors which comes in a custom object.
Custom objects:
1. contrastive loss
2. embedding layer: where we are finding euclidean_distance.

Saving model:
for saving model it is straight forward <model_name>.save("siamese_contrastive.h5")

Loading model:
Here the good part will come model will not load directly here because it doesn't have an understanding of two things one is your custom layer and 2nd is your loss.

model = tf.keras.models.load_model('siamese_contrastive.h5', custom_objects={ })

In the custom object mentioned above, you have to provide the definition of those two objects.
After that, it will accept your model and it will run separately at inferencing time.

still figuring out how.
Have a look at my implementation let me know if you still have any doubts
link: https://github.com/anukash/Face_recognition_siamese_network

**Comment by SuryanarayanaY** [comment]
Hi @sherlock666 ,

We need to add `@keras.saving.register_keras_serializable()` decorator to the custom_object in the model to make it 
serializable. I have done the changes for custom loss function as below.

```
@keras.saving.register_keras_serializable(name="loss")
def loss(margin=1):
    """Provides 'contrastive_loss' an enclosing scope with variable 'margin'.

    Arguments:
        margin: Integer, defines the baseline for distance for which pairs
                should be classified as dissimilar. - (default is 1).

    Returns:
        'contrastive_loss' function with data ('margin') attached.
    """

    # Contrastive loss = mean( (1-true_value) * square(prediction) +
    #                         true_value * square( max(margin-prediction, 0) ))
    @keras.saving.register_keras_serializable(name="contrastive_loss")
    def contrastive_loss(y_true, y_pred):
        """Calculates the contrastive loss.

        Arguments:
            y_true: List of labels, each label is of type float32.
            y_pred: List of predictions of same length as of y_true,
                    each label is of type float32.

        Returns:
            A tensor containing contrastive loss as floating point value.
        """

        square_pred = tf.math.square(y_pred)
        margin_square = tf.math.square(tf.math.maximum(margin - (y_pred), 0))
        return tf.math.reduce_mean(
            (1 - y_true) * square_pred + (y_true) * margin_square
        )

    return contrastive_loss
```

Additionally we need to pass safe_mode=False to reload the model since our loss function is a Python function.

`reloaded_model = keras.models.load_model('siamese.keras',safe_mode=False)
`

With above changes I have tested the tutorial with TF2.13v and .keras save format and the model is able to save and load properly. I also run `tf.math.reduce_all(predictions) == tf.math.reduce_all(reloaded_model_predictions)` which returns True that indicates reloading of the model also success.

Please find the attached [gist](https://colab.sandbox.google.com/gist/SuryanarayanaY/c1bfeff63deb91822885da2ce83aa330/kerasio_668-siamese_contrastive.ipynb) for all above modifications and outputs verifications. 

Hope this will resolve your query. Please verify and feel free to close the issue if resolved.

Thanks!
